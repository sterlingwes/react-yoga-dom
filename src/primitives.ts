import yoga from 'yoga-layout';
import { UnreachableCaseError } from './utils';

export enum FlexJustifyValue {
  flexStart = 'flex-start',
  flexEnd = 'flex-end',
  center = 'center',
  spaceBetween = 'space-between',
  spaceAround = 'space-around',
  spaceEvenly = 'space-evenly',
}

export enum FlexAlignValue {
  flexStart = 'flex-start',
  flexEnd = 'flex-end',
  center = 'center',
  stretch = 'stretch',
  baseline = 'baseline',
}

export enum LayoutDirection {
  inherit = 'inherit',
  ltr = 'ltr',
  rtl = 'rtl',
}

export enum FlexDisplay {
  none = 'none',
  flex = 'flex',
}

export enum FlexDirection {
  row = 'row',
  rowReverse = 'row-reverse',
  column = 'column',
  columnReverse = 'column-reverse',
}

export enum FlexWrap {
  wrap = 'wrap',
  nowrap = 'nowrap',
}

export enum OverflowValue {
  visible = 'visible',
  hidden = 'hidden',
  scroll = 'scroll',
}

export enum PositionValue {
  absolute = 'absolute',
  relative = 'relative',
}

type LayoutPositionT = ReturnType<typeof LayoutPosition>;
type PositionValues = { [key in 'top' | 'left' | 'right' | 'bottom']: LayoutValue };

export type LayoutValue = number | string;

// ltr / rtl props not implemented; here for reference
// also not supporting some percentage (string) dimension values yet
// (minWidthPercent etc)
export interface RNStyleT {
  alignContent?: FlexJustifyValue;
  alignItems?: FlexAlignValue;
  alignSelf?: FlexAlignValue;
  aspectRatio?: number;
  borderBottomWidth?: number;
  borderEndWidth?: number;
  borderLeftWidth?: number;
  borderRightWidth?: number;
  borderStartWidth?: number;
  borderTopWidth?: number;
  borderWidth?: number;
  bottom?: LayoutValue;
  direction?: LayoutDirection;
  display?: FlexDisplay;
  end?: LayoutValue;
  flex?: number;
  flexBasis?: LayoutValue;
  flexDirection?: FlexDirection;
  flexGrow?: number;
  flexShrink?: number;
  flexWrap?: FlexWrap;
  height?: string | number;
  justifyContent?: FlexJustifyValue;
  left?: LayoutValue;
  margin?: LayoutValue;
  marginBottom?: LayoutValue;
  marginEnd?: LayoutValue;
  marginHorizontal?: LayoutValue;
  marginLeft?: LayoutValue;
  marginRight?: LayoutValue;
  marginStart?: LayoutValue;
  marginTop?: LayoutValue;
  marginVertical?: LayoutValue;
  maxHeight?: LayoutValue;
  maxWidth?: LayoutValue;
  minHeight?: LayoutValue;
  minWidth?: LayoutValue;
  overflow?: OverflowValue;
  padding?: LayoutValue;
  paddingBottom?: LayoutValue;
  paddingEnd?: LayoutValue;
  paddingHorizontal?: LayoutValue;
  paddingLeft?: LayoutValue;
  paddingRight?: LayoutValue;
  paddingStart?: LayoutValue;
  paddingTop?: LayoutValue;
  paddingVertical?: LayoutValue;
  position?: PositionValue;
  right?: LayoutValue;
  start?: LayoutValue;
  top?: LayoutValue;
  width?: LayoutValue;
  zIndex?: number;
}

const selectConsistentStyles = (style: RNStyleT) => {
  const { width, height } = style;
  return {
    width,
    height,
  };
};

const resolveFlexBasis = (style: RNStyleT) => {
  if (style.flexBasis) return style.flexBasis;
  if (style.flex && style.flex > 0) return 0;
  return 'auto';
};

const resolveFlexGrow = (style: RNStyleT) => {
  if (style.flexGrow) return style.flexGrow;
  if (style.flex && style.flex > 0) return style.flex;
  return 0;
};

const resolveFlexShrink = (style: RNStyleT) => {
  if (style.flexShrink) return style.flexShrink;
  if (style.flex && style.flex < 0) return -style.flex;
  return 0;
};

const convertAlignValue = (value?: FlexJustifyValue | FlexAlignValue) => {
  switch (value) {
    case undefined:
      return yoga.ALIGN_AUTO;
    case FlexAlignValue.baseline:
      return yoga.ALIGN_BASELINE;
    case FlexAlignValue.center:
    case FlexJustifyValue.center:
      return yoga.ALIGN_CENTER;
    case FlexAlignValue.flexEnd:
    case FlexJustifyValue.flexEnd:
      return yoga.ALIGN_FLEX_END;
    case FlexAlignValue.flexStart:
    case FlexJustifyValue.flexStart:
      return yoga.ALIGN_FLEX_START;
    case FlexJustifyValue.spaceAround:
      return yoga.ALIGN_SPACE_AROUND;
    case FlexJustifyValue.spaceBetween:
      return yoga.ALIGN_SPACE_BETWEEN;
    case FlexJustifyValue.spaceEvenly:
      return yoga.JUSTIFY_SPACE_EVENLY;
    case FlexAlignValue.stretch:
      return yoga.ALIGN_STRETCH;
    default:
      throw new UnreachableCaseError(value);
  }
};

const convertAlignContent = (style: RNStyleT) => {
  const { alignContent } = style;
  return convertAlignValue(alignContent);
};

const convertJustifyContent = (style: RNStyleT) => {
  const { justifyContent } = style;
  if (!justifyContent) return yoga.JUSTIFY_FLEX_START;
  switch (justifyContent) {
    case FlexJustifyValue.center:
      return yoga.JUSTIFY_CENTER;
    case FlexJustifyValue.flexEnd:
      return yoga.JUSTIFY_FLEX_END;
    case FlexJustifyValue.flexStart:
      return yoga.JUSTIFY_FLEX_START;
    case FlexJustifyValue.spaceAround:
      return yoga.JUSTIFY_SPACE_AROUND;
    case FlexJustifyValue.spaceBetween:
      return yoga.JUSTIFY_SPACE_BETWEEN;
    case FlexJustifyValue.spaceEvenly:
      return yoga.JUSTIFY_SPACE_EVENLY;
    default:
      throw new UnreachableCaseError(justifyContent);
  }
};

const convertAlignItems = (style: RNStyleT) => {
  const { alignItems } = style;
  return convertAlignValue(alignItems);
};

const convertAlignSelf = (style: RNStyleT) => {
  const { alignSelf } = style;
  return convertAlignValue(alignSelf);
};

const convertFlexDirection = (style: RNStyleT) => {
  const { flexDirection } = style;
  switch (flexDirection) {
    case undefined:
    case FlexDirection.column:
      return yoga.FLEX_DIRECTION_COLUMN;
    case FlexDirection.columnReverse:
      return yoga.FLEX_DIRECTION_COLUMN_REVERSE;
    case FlexDirection.row:
      return yoga.FLEX_DIRECTION_ROW;
    case FlexDirection.rowReverse:
      return yoga.FLEX_DIRECTION_ROW_REVERSE;
    default:
      throw new UnreachableCaseError(flexDirection);
  }
};

const convertPositionType = (style: RNStyleT) => {
  const { position } = style;
  switch (position) {
    case undefined:
    case PositionValue.relative:
      return yoga.POSITION_TYPE_RELATIVE;
    case PositionValue.absolute:
      return yoga.POSITION_TYPE_ABSOLUTE;
    default:
      throw new UnreachableCaseError(position);
  }
};

const convertDimension = (value?: LayoutValue) => {
  if (!value) return 'auto';
  return value;
};

const convertFlexWrap = (style: RNStyleT) => {
  const { flexWrap } = style;
  switch (flexWrap) {
    case undefined:
    case FlexWrap.nowrap:
      return yoga.WRAP_NO_WRAP;
    case FlexWrap.wrap:
      return yoga.WRAP_WRAP;
    default:
      throw new UnreachableCaseError(flexWrap);
  }
};

const convertBordersToPosition = (style: RNStyleT): LayoutPositionT => {
  const {
    borderBottomWidth,
    borderLeftWidth,
    borderRightWidth,
    borderTopWidth,
    borderWidth,
  } = style;

  return LayoutPosition({
    top: borderTopWidth || borderWidth || 0,
    left: borderLeftWidth || borderWidth || 0,
    right: borderRightWidth || borderWidth || 0,
    bottom: borderBottomWidth || borderWidth || 0,
  });
};

const convertPaddingToPosition = (style: RNStyleT): LayoutPositionT => {
  const {
    padding,
    paddingBottom,
    paddingHorizontal,
    paddingLeft,
    paddingRight,
    paddingTop,
    paddingVertical,
  } = style;

  return LayoutPosition({
    top: paddingTop || paddingVertical || padding || 0,
    left: paddingLeft || paddingHorizontal || padding || 0,
    right: paddingRight || paddingHorizontal || padding || 0,
    bottom: paddingBottom || paddingVertical || padding || 0,
  });
};

const convertMarginToPosition = (style: RNStyleT): LayoutPositionT => {
  const {
    margin,
    marginBottom,
    marginHorizontal,
    marginLeft,
    marginRight,
    marginTop,
    marginVertical,
  } = style;

  return LayoutPosition({
    top: marginTop || marginVertical || margin || 0,
    left: marginLeft || marginHorizontal || margin || 0,
    right: marginRight || marginHorizontal || margin || 0,
    bottom: marginBottom || marginVertical || margin || 0,
  });
};

const convertPosition = (style: RNStyleT): PositionValues => {
  const { top, left, right, bottom } = style;
  return {
    top: top || NaN,
    left: left || NaN,
    right: right || NaN,
    bottom: bottom || NaN,
  };
};

export const createLayout = (style: RNStyleT): LayoutT => {
  const flowThrough = selectConsistentStyles(style);
  return {
    ...flowThrough,
    flexBasis: resolveFlexBasis(style),
    flexGrow: resolveFlexGrow(style),
    flexShrink: resolveFlexShrink(style),
    flexWrap: convertFlexWrap(style),
    alignContent: convertAlignContent(style),
    justifyContent: convertJustifyContent(style),
    alignItems: convertAlignItems(style),
    alignSelf: convertAlignSelf(style),
    flexDirection: convertFlexDirection(style),
    positionType: convertPositionType(style),
    width: convertDimension(style.width),
    height: convertDimension(style.height),
    padding: convertPaddingToPosition(style),
    margin: convertMarginToPosition(style),
    border: convertBordersToPosition(style),
    position: convertPosition(style),
  };
};

export const LayoutPosition = (positions = {}) => ({
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  ...positions,
});

export interface LayoutT {
  width?: LayoutValue;
  height?: LayoutValue;
  justifyContent?: yoga.YogaJustifyContent;
  alignItems?: yoga.YogaAlign;
  alignSelf?: yoga.YogaAlign;
  alignContent?: yoga.YogaAlign;
  flexDirection?: number;
  padding?: LayoutPositionT;
  margin?: LayoutPositionT;
  border?: LayoutPositionT;
  position?: PositionValues;
  positionType?: yoga.YogaPositionType;
  flexWrap?: yoga.YogaFlexWrap;
  flexBasis?: LayoutValue;
  flexGrow?: number;
  flexShrink?: number;
  children?: Array<LayoutT>;
  aspectRatio?: number;
  minWidth?: number;
  maxWidth?: number;
  minHeight?: number;
  maxHeight?: number;
}
