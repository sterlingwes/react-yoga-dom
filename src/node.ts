import yoga, { Node, YogaNode } from 'yoga-layout';
import { LayoutT, LayoutPosition, LayoutValue } from './primitives';
import { UnreachableCaseError } from './utils';

type SettableProperty =
  | 'width'
  | 'height'
  | 'minWidth'
  | 'maxWidth'
  | 'minHeight'
  | 'maxHeight'
  | 'justifyContent'
  | 'alignItems'
  | 'alignSelf'
  | 'alignContent'
  | 'flexGrow'
  | 'flexShrink'
  | 'positionType'
  | 'aspectRatio'
  | 'flexWrap'
  | 'flexDirection';

const SETTABLE_PROPERTIES: SettableProperty[] = [
  'width',
  'height',
  'minWidth',
  'maxWidth',
  'minHeight',
  'maxHeight',
  'justifyContent',
  'alignItems',
  'alignSelf',
  'alignContent',
  'flexGrow',
  'flexShrink',
  'positionType',
  'aspectRatio',
  'flexWrap',
  'flexDirection',
];

type LayoutType = 'padding' | 'margin' | 'position' | 'border';
const LAYOUT_TYPE: LayoutType[] = ['padding', 'margin', 'position', 'border'];

type LayoutEdge = 'top' | 'left' | 'bottom' | 'right';
const LAYOUT_EDGE: LayoutEdge[] = ['left', 'top', 'right', 'bottom'];
const EDGE_MAP: { [key in LayoutEdge]: yoga.YogaEdge } = {
  top: yoga.EDGE_TOP,
  left: yoga.EDGE_LEFT,
  bottom: yoga.EDGE_BOTTOM,
  right: yoga.EDGE_RIGHT,
};

const isPercent = (value: LayoutValue) => typeof value === 'string' && value !== 'auto';
const asPercent = (value: LayoutValue) => (typeof value === 'string' ? parseFloat(value) : value);

export const createYogaNodes = (layout: LayoutT): YogaNode => {
  const root = Node.create();

  SETTABLE_PROPERTIES.forEach(key => {
    switch (key) {
      case 'width':
        if (layout.width) {
          if (isPercent(layout.width)) {
            root.setWidthPercent(asPercent(layout.width));
            break;
          }
          root.setWidth(layout.width);
        }
        break;
      case 'height':
        if (layout.height) {
          if (isPercent(layout.height)) {
            root.setHeightPercent(asPercent(layout.height));
            break;
          }
          root.setHeight(layout.height);
        }
        break;
      case 'minWidth':
        layout.minWidth && root.setMinWidth(layout.minWidth);
        break;
      case 'maxWidth':
        layout.maxWidth && root.setMaxWidth(layout.maxWidth);
        break;
      case 'minHeight':
        layout.minHeight && root.setMinHeight(layout.minHeight);
        break;
      case 'maxHeight':
        layout.maxHeight && root.setMaxWidth(layout.maxHeight);
        break;
      case 'justifyContent':
        layout.justifyContent && root.setJustifyContent(layout.justifyContent);
        break;
      case 'alignItems':
        layout.alignItems && root.setAlignItems(layout.alignItems);
        break;
      case 'alignSelf':
        layout.alignSelf && root.setAlignSelf(layout.alignSelf);
        break;
      case 'alignContent':
        layout.alignContent && root.setAlignContent(layout.alignContent);
        break;
      case 'flexGrow':
        layout.flexGrow && root.setFlexGrow(layout.flexGrow);
        break;
      case 'flexShrink':
        layout.flexShrink && root.setFlexShrink(layout.flexShrink);
        break;
      case 'positionType':
        layout.positionType && root.setPositionType(layout.positionType);
        break;
      case 'aspectRatio':
        layout.aspectRatio && root.setAspectRatio(layout.aspectRatio);
        break;
      case 'flexWrap':
        layout.flexWrap && root.setFlexWrap(layout.flexWrap);
        break;
      case 'flexDirection':
        break;
      default:
        throw new UnreachableCaseError(key);
    }
  });

  LAYOUT_TYPE.forEach(type => {
    LAYOUT_EDGE.forEach(edge => {
      const edgeVal = EDGE_MAP[edge];
      const edgeKey = LAYOUT_EDGE[edgeVal];

      switch (type) {
        case 'border':
          layout.border && root.setBorder(edgeVal, layout.border[edgeKey]);
          break;
        case 'margin':
          if (layout.margin) {
            const marginVal = layout.margin[edgeKey];
            if (isPercent(marginVal)) {
              root.setMarginPercent(edgeVal, asPercent(marginVal));
              break;
            }
            root.setMargin(edgeVal, layout.margin[edgeKey]);
          }
          break;
        case 'padding':
          if (layout.padding) {
            const paddingval = layout.padding[edgeKey];
            if (isPercent(paddingval)) {
              root.setPaddingPercent(edgeVal, asPercent(paddingval));
              break;
            }
            root.setPadding(edgeVal, layout.padding[edgeKey]);
          }
          break;
        case 'position':
          if (layout.position) {
            const positionVal = layout.position[edgeKey];
            if (isPercent(positionVal)) {
              root.setPositionPercent(edgeVal, asPercent(positionVal));
              break;
            }
            root.setPosition(edgeVal, positionVal);
          }
          break;
        default:
          throw new UnreachableCaseError(type);
      }
    });
  });

  root.setDisplay(yoga.DISPLAY_FLEX);

  (layout.children || []).map(createYogaNodes).forEach((node, i) => {
    root.insertChild(node, i);
  });
  return root;
};
