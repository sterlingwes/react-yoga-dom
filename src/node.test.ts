import yoga, { YogaNode } from 'yoga-layout';
import { createLayout, LayoutT, PositionValue } from './primitives';
import { createYogaNodes } from './node';

type IteratorT = (value: any, index: number, items: any[]) => unknown;
const mapTimes = (n: number, iterator: IteratorT) => [...Array(n)].map(iterator);

describe('createYogaNodes', () => {
  let layout: LayoutT;

  beforeEach(() => {
    layout = createLayout({
      width: 500,
      height: 500,
    });
  });

  const calculateLayout = () => {
    const node = createYogaNodes(layout);
    node.calculateLayout(layout.width as number, layout.height as number);
    const childCount = node.getChildCount();
    const children = mapTimes(childCount, (_, i) => node.getChild(i).getComputedLayout());
    return { node, children };
  };

  it('should offset children with absolute top', () => {
    layout.children = [...Array(3)].map(() => createLayout({ height: 20 }));
    const { children } = calculateLayout();
    expect(children).toMatchSnapshot();
  });

  it('should account for margin values in layout', () => {
    layout.children = [...Array(3)].map(() => createLayout({ height: 20, marginBottom: 1 }));
    const { children } = calculateLayout();
    expect(children).toMatchSnapshot();
  });

  it('should position relative by default', () => {
    layout.children = [...Array(3)].map(() => createLayout({ height: 20, top: 1 }));
    const { children } = calculateLayout();
    expect(children).toMatchSnapshot();
  });

  it('should allow for absolute positioning', () => {
    layout.children = [...Array(3)].map(() =>
      createLayout({ height: 20, top: 1, position: PositionValue.absolute }),
    );
    const { children } = calculateLayout();
    expect(children).toMatchSnapshot();
  });

  it('should allow for percentage positioning', () => {
    layout.children = [...Array(3)].map(() =>
      createLayout({ height: 20, top: '30%', position: PositionValue.absolute }),
    );
    const { children } = calculateLayout();
    expect(children).toMatchSnapshot();
  });
});
