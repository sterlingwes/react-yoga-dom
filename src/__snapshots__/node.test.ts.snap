// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createYogaNodes should account for margin values in layout 1`] = `
Array [
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 0,
    "width": 500,
  },
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 21,
    "width": 500,
  },
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 42,
    "width": 500,
  },
]
`;

exports[`createYogaNodes should allow for absolute positioning 1`] = `
Array [
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 1,
    "width": 0,
  },
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 1,
    "width": 0,
  },
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 1,
    "width": 0,
  },
]
`;

exports[`createYogaNodes should allow for percentage positioning 1`] = `
Array [
  Layout {
    "bottom": 150,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 150,
    "width": 0,
  },
  Layout {
    "bottom": 150,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 150,
    "width": 0,
  },
  Layout {
    "bottom": 150,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 150,
    "width": 0,
  },
]
`;

exports[`createYogaNodes should offset children with absolute top 1`] = `
Array [
  Layout {
    "bottom": 0,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 0,
    "width": 500,
  },
  Layout {
    "bottom": 0,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 20,
    "width": 500,
  },
  Layout {
    "bottom": 0,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 40,
    "width": 500,
  },
]
`;

exports[`createYogaNodes should position relative by default 1`] = `
Array [
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 1,
    "width": 500,
  },
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 21,
    "width": 500,
  },
  Layout {
    "bottom": 1,
    "height": 20,
    "left": 0,
    "right": 0,
    "top": 41,
    "width": 500,
  },
]
`;
